paritytech/parity-ethereum
# [Trace] Distinguish between `create` and `create2`

Closes #10922 and implements the approach suggested by @tomusdrw.

This PR introduces a new field on the *Create* struct in the *trace* module. This allows differentiating between create and create2 contract deployments. The information comes from the *exec_instruction* method inside the *ethcore* interpreter module (there we have access to the low level OpCode).

It's then passed on via the *ActionType* enum (previously *CallType*) on the *ActionParams* struct, and eventually translated into the trace object.

It looks like at the moment `CREATE` opcodes are stored as `CallType::None`. I therefore reused this subtype (persisted as `0u32`) and added a new one for CREATE2.

### Test plan

1. Checkout master (not this PR), compile and run private test chain
```
./target/debug/parity --config dev --tracing=on --unlock="0x00a329c0648769a73afac7f9381e08fb43dbea72" --password=./password
```
2. Checkout [this test repo](https://github.com/fleupold/solidity-create2-example) with a simple deploy2 factory
    - `npm ci`
    - `make deploy`
    - `make test` (capture block number that is printed)
3. Get trace of that block
```
curl -w "\n" -H "Content-Type: application/json" -X POST --data '{"jsonrpc":"2.0","method":"trace_replayBlockTransactions","params":["<block_number>",["trace"]],"id":67}' http://localhost:8545 | python -m json.tool
```
4. See no creation type in json output
5. Checkout this PR, build and run on same private chain
6. Rerun the rpc call with the same block number
7. Now observe `"create_type": "create2",` in output
8. The trace of `block_number - 1` should have a `"create_type": "create",`
9. Same will work if you run `make test` again
